Value-set analysis: templatise and virtualise to facilitate customisation

smowton opened the pull request:
This provides sufficient hooks in value-set analysis, a CBMC core pointer-analysis, to permit downstream products such as the security analyser to create derivatives that share most of its logic.

VSA itself is currently unused, as revealed by the need to add support for DEAD statements before it can be used; this also adds unit tests to check its basic functionality has not broken in the meantime.

VSA is based on `static_analysist` and so will need to be ported to `ait` in the future if we wish to continue using it; however this should be orthogonal to permitting it to be customised like this.

smowton committed:
Templatize and virtualize value-set analysis

This templates value-set-analysis, so that it can be subclassed
providing a value-set extension, and virtualizes value-set so it
can similarly be extended.

smowton committed:
Add unit tests for value-set-analysis customisation

smowton committed:
Value-set analysis: ignore DEAD statements

We could be smarter with these; this at least enables us to process
current standard GOTO programs.

reuk commented on src/pointer-analysis/value_set.h:
```diff
@@ -211,11 +211,11 @@ class value_sett
     const namespacet &ns);
 
-  void apply_code(
+  virtual void apply_code(
```
In general `public virtual` methods in non-abstract classes are a code smell, as they have two many responsibilities. They specify both the public interface, and a customization point for derived classes (which is an implementation detail). Ideally, the two responsibilities should be decoupled by providing a `private virtual` method which is called by a `public` (non-virtual) method. Then, if interface requirements change, the interface can be adjusted independently of the implementation (and vice versa).

I'm particularly concerned about retroactively making this method virtual. Pre-existing clients which always expect this method to do the same thing could be broken when derived classes change its behavior.

reuk commented on src/pointer-analysis/value_set.h:
```diff
@@ -242,28 +242,28 @@ class value_sett
     const namespacet &ns) const;
 
 protected:
```
Prefer to make these `private` if at all possible.

smowton replied:
The virtual override points definitely need defer-to-base I'm afraid

reuk commented on src/pointer-analysis/value_set_analysis.h:
```diff
@@ -14,32 +14,90 @@ Author: Daniel Kroening, kroening@kroening.com
 class xmlt;
 
-class value_set_analysist:
+template<class Value_Sett>
```
Plz no camel snake case. `ValueSet`?

thk123 replied:
I don't think adopting a different naming convention is a good way of disambiguating it from the existing `value_sett`. Perhaps `selected_value_sett`? Perhaps there is a better term. `value_set_domaint`?

reuk commented on src/pointer-analysis/value_set_analysis.h:
```diff
@@ -14,32 +14,90 @@ Author: Daniel Kroening, kroening@kroening.com
+  typedef static_analysist<domaint> baset;
+  typedef typename baset::locationt locationt;
+
+  explicit value_set_analysis_baset(const namespacet &_ns):baset(_ns)
```
No leading underscores on identifiers.

reuk commented on src/pointer-analysis/value_set_analysis.h:
```diff
@@ -14,32 +14,90 @@ Author: Daniel Kroening, kroening@kroening.com
-  virtual void initialize(const goto_functionst &goto_functions);
+  void initialize(const goto_programt &goto_program)
+  {
+    baset::initialize(goto_program);
```
Overridden methods calling base class methods is a code smell. It implies that the customization points have been chosen incorrectly. This should probably be refactored.

reuk commented on src/pointer-analysis/value_set_analysis.h:
```diff
@@ -14,32 +14,90 @@ Author: Daniel Kroening, kroening@kroening.com
+  {
+    baset::initialize(goto_program);
+  }
+  void initialize(const goto_functionst &goto_functions)
```
I think these methods are missing `override`s too.

reuk commented on src/pointer-analysis/value_set_analysis.h:
```diff
@@ -14,32 +14,90 @@ Author: Daniel Kroening, kroening@kroening.com
+        var.new_element("identifier").data=
+          id2string(v_it->first);
+
+#if 0
```
If this is never used it should probably be removed.

smowton replied:
Shrug, that goes for the entire XML interface. I don't know who (if anyone) uses it and if they use this sometimes, so I'm leaving it alone.

reuk commented on src/pointer-analysis/value_set_domain.h:
```diff
@@ -17,10 +17,11 @@ Author: Daniel Kroening, kroening@kroening.com
 {
 public:
-  value_sett value_set;
+  Value_Sett value_set;
```
More snake-camel case

reuk commented on src/pointer-analysis/value_set_domain.h:
```diff
@@ -44,18 +45,59 @@ class value_set_domaint:public domain_baset
+      break;
+    }
+
+    case RETURN:
```
Add a comment to make it clear that fall-through is desired.

reuk commented on src/pointer-analysis/value_set.h:
```diff
@@ -242,28 +242,28 @@ class value_sett
 
 protected:
-  void get_value_set_rec(
+  virtual void get_value_set_rec(
```
Again, I find it extremely unlikely that these methods are already suitable customization points. It might be better to find a more focused group of custom behaviors which can be given private virtual interfaces, and called from these (still non-virtual) methods.

martin-cs commented:
On Fri, 2017-09-22 at 13:26 +0000, Chris Smowton wrote:
> This provides sufficient hooks in value-set analysis, a CBMC core pointer-analysis, to permit downstream products such as the security analyser to create derivatives that share most of its logic.

Sounds good.

> VSA itself is currently unused, as revealed by the need to add support for DEAD statements before it can be used; this also adds unit tests to check its basic functionality has not broken in the meantime.

I don't quite recall the history but it was either written for CBMC but
we used a slightly different version or it was used in other tools;
possibly Wolverine...

> VSA is based on `static_analysist` and so will need to be ported to `ait` in the future if we wish to continue using it;

Yes; this was the last part of the "we should probably only have one
general abstract interpretation framework".

>  however this should be orthogonal to permitting it to be customised like this.

Yes.  There is also quite a bit of duplicate code in the various value
analysis variants; if you are refactoring it might be worth having a
look at that (the duplication was one of the motivations behind the
variable precision domain stuff).  Do be a little careful though because
IIRC there are some subtle hacks in this code to handle C pointer
weirdness.

Longer term there may be a question over whether configurations of the
variable precision domain might replace the functionality of
pointer_analysis/ but even if it does, it will be a while before this
code is removed.

HTH

thk123 commented on src/pointer-analysis/show_value_sets.h:
```diff
@@ -12,10 +12,10 @@ Author: Daniel Kroening, kroening@kroening.com
 #ifndef CPROVER_POINTER_ANALYSIS_SHOW_VALUE_SETS_H
 #define CPROVER_POINTER_ANALYSIS_SHOW_VALUE_SETS_H
 
+#include <pointer-analysis/value_set_analysis.h>
```
Why swap this to a full include when forward declaration was sufficient?

smowton replied:
`value_set_analysist` is a template specialisation now; `class value_set_analysist` won't work as the forward-decl will clash with the actual definition. There might be a way around this but I don't know it.

thk123 commented on src/pointer-analysis/value_set_analysis.h:
```diff
@@ -14,32 +14,90 @@ Author: Daniel Kroening, kroening@kroening.com
+  {
+    source_locationt previous_location;
+
+    forall_goto_program_instructions(i_it, goto_program)
```
Nit: while in the area a range based for might be better here?

smowton replied:
True, re: this and below though I'll stick to just moving the code without many changes and perhaps clean up later.

thk123 commented on src/pointer-analysis/value_set_analysis.h:
```diff
@@ -14,32 +14,90 @@ Author: Daniel Kroening, kroening@kroening.com
+      xmlt &i=dest.new_element("instruction");
+      i.new_element()=::xml(location);
+
+      for(value_sett::valuest::const_iterator
```
Nit: the formatting on this for loop is a bit hard to follow. You might consider using `auto` to get each statement on one line or at least aligning each statement

thk123 commented on src/pointer-analysis/value_set_analysis.h:
```diff
@@ -48,10 +106,12 @@ class value_set_analysist:
   }
 };
 
+typedef value_set_analysis_baset<value_sett> value_set_analysist;
```
This seems like a weird place to introduce a typedef? If really needed in the header file outside of the class, perhaps put it above the class?

thk123 commented on unit/CustomVSATest.java:
```diff
@@ -0,0 +1,51 @@
+
+public class CustomVSATest {
```
Could this go in the same folder as the test to avoid cluttering `unit`

thk123 commented on unit/pointer-analysis/custom_value_set_analysis.cpp:
```diff
@@ -0,0 +1,387 @@
+    // constructor callsites confusing the results we're trying to check:
+    goto_function_inline(goto_model, TEST_FUNCTION_NAME, null_output);
+
+    REQUIRE(!model_init_failed);
```
Any reason not to do this REQUIRE immediately after `initalize_goto_model`?

thk123 reviewed (commented):
This looks good - just a few minor corrections. I also agree with @reuk that the interface could probably be cleaned up so the public methods aren't virtual. I will read up on Value Set Analysis so that I can validate that the tests covers suitable cases - but they look very clean!

thk123 commented on unit/pointer-analysis/custom_value_set_analysis.cpp:
```diff
@@ -0,0 +1,387 @@
+/// * Adds an ID_unknown to the value of variable "maybe_unknown", and
+///     to the possible referees of fields named unknown_field_ref and
+///     variables named unknown_global_ref
+class custom_value_sett:public value_sett
```
You might consider pulling this class out into a separate file with just the above documentation left in the test file since it isn't directly relevant to the test. Definitely not a requirement, so do disregard if you disagree.

thk123 reviewed (commented):
The tests seem good. I'm tempted to say the methods that you don't provide a custom implementation for probably shouldn't be virtual

martin-cs commented:
> Shrug, that goes for the entire XML interface. I don't know who (if anyone) uses it and if they use this sometimes, so I'm leaving it alone. 

The XML interface was developed for a customer (I forget exactly who)
and I think there are several people who use it. Any decision to
depreciate it should be done across the whole code-base.

smowton commented:
@reuk @thk123 all changes applied except as noted.

thk123 commented on src/pointer-analysis/value_set_analysis.h:
```diff
@@ -14,32 +14,80 @@ Author: Daniel Kroening, kroening@kroening.com
 class xmlt;
 
-class value_set_analysist:
+template<class DomainValueSet>
```
Still think looks a bit awkward to have this non-conforming to name convention (pointing out in case this isn't intentional, not to start a flame war :stuck_out_tongue:)

thk123 reviewed (approved):
The code and tests look great, I would feel more conformable if someone who was already familiar with the behaviour of these analyses checked the tests were sufficient.

smowton commented:
Changed names again to VST, believe everything is applied now

reuk commented on src/pointer-analysis/value_set.h:
```diff
@@ -287,6 +288,54 @@ class value_sett
+
+  /// Expression simplification function; by default, plain old
+  /// util/simplify_expr, but can be customised by subclass.
+  expr_simplifiert simplifier;
```
Make private, and provide a protected method which allows this to be called (but nothing else)

reuk commented on src/pointer-analysis/value_set.h:
```diff
@@ -24,8 +24,20 @@ class namespacet;
+
+  explicit value_sett(expr_simplifiert simplifier):
+    location_number(0),
+    simplifier(simplifier)
```
Should `std::move` the simplifier

reuk commented on unit/pointer-analysis/custom_value_set_analysis.cpp:
```diff
@@ -0,0 +1,336 @@
+#include <java_bytecode/java_types.h>
+#include <pointer-analysis/value_set_analysis.h>
+
+/// Counts calls to our custom simplifier, registered above:
```
Below...

reuk commented on unit/pointer-analysis/custom_value_set_analysis.cpp:
```diff
@@ -0,0 +1,336 @@
+
+#define TEST_PREFIX "java::CustomVSATest."
+#define TEST_FUNCTION_NAME TEST_PREFIX "test:()V"
+#define TEST_LOCAL_PREFIX TEST_FUNCTION_NAME "::"
```
You could use `static constexpr const char *` here and avoid the evil preprocessor

smowton replied:
AFAIK I can't concatenate them as is done here, so I'd prefer to leave as-is

reuk commented on src/pointer-analysis/value_set_domain.h:
```diff
@@ -44,17 +45,59 @@ class value_set_domaint:public domain_baset
     value_setst::valuest &dest)
   {
-    value_set.get_reference_set(expr, dest, ns);
+    value_set.read_reference_set(expr, dest, ns);
```
In general, if you are specifying interface requirements on a templated item, use non-member rather than member functions (i.e. `read_reference_set(value_set, expr, dest, ns)`). This allows pre-existing types to be used, without altering their definitions, as the template requirements can be fulfilled by providing external, non-member functions. If necessary, these can even be local to a translation unit, allowing template clients in different locations to use different implementations of these wrapper/interface functions.

smowton replied:
Don't follow this, could I have an example?

reuk replied:
Let's say I have a type that already exists - maybe it's a `std::vector` or something where we're not able to change the implementation. Now, if you were to use this type as your template parameter, you would probably find that it doesn't have the right member functions to satisfy the interface implicitly specified by the template class. The problem is that there's nothing you can do now - you can't modify the class definition to add the right methods.

If you use non-member functions inside your template, you have more flexibility to define appropriate implementations of these functions without modifying the original class. This also gives more flexibility when refactoring. At the moment, if you changed a method in one of your classes that's used as a `VST`, you would have to go and modify this template, too. However, if you use non-member functions you can just modify the non-member passthrough function to call the modified member function (one change), without having to make changes throughout the template class (potentially several changes).

This is about separating the general interface for a class (its member functions) from the specific interface required for a particular template class (non-member functions).

smowton replied:
OK, done this this time. In general I think this sort of layer of indirection should be introduced on demand though, as the case where we need it is fairly unlikely (maybe 10% chance of needing this over the life of the code, considering the template parameters are internal code, and the cost of introducing the indirection scales with the size of the implied interface, not the number of subclasses already providing it.

smowton commented:
Comments 1-3 done, 4 declined, 5 I don't understand, please give an example?

smowton committed:
Indirect value_set_domaint -> value_sett operations via non-member functions

These provide an oportunity for future implementers of value-sets not to have to
conform to the same public interface as value_sett, e.g. to provide some type they
can't or don't want to augment and provide non-member function specialisations
instead.

smowton commented:
Change 5 now applied too.

polgreen commented:
Hi, 
It would be helpful if in future, when large-ish changes like this are introduced, the old api could be kept in develop in addition to the new one at least for a short time, please. Then people who depend on the existing api have a bit of time to alter their code to fit the new api, whilst still being able to rebase against develop to pull in other new changes.
Thanks!

smowton commented:
1. Sorry, didn't think anybody was using this code except for goto-instrument.
2. What API break? `value_set_analysist` and `value_sett` *should* be exactly as-was
3. Daniel has requested this reverted, which will happen once CI has passed

martin-cs commented:
On Mon, 2017-10-16 at 01:28 -0700, Chris Smowton wrote:
> 1. Sorry, didn't think anybody was using this code except for goto-instrument.
It should be used in symex and in the dependency graph / slicer at a
minimum.  As a generally point, we can't entirely assume that something
is unused just because it is not used in the main CBMC repository.

Also, where possible, introducing the new API, depreciating the old one
and then removing it later is A Good Thing.  See the ai.h /
static_analysis.h divide.

polgreen commented:
I'm using value_set_domaint, which has changed quite a bit. I've fixed my stuff now, just it would be nice if there's a bit of a period of overlap for future changes like this.

smowton commented:
Ah ok, so would my writing typedef value_set_domaint templated_value_set_domaint<value_sett> have prevented problems, or were there further problems with value_set_domaint's internals?

polgreen commented:
You would also need to keep the names of the functions within templated_value_set_domaint<value_sett> the same as the functions within value_set_domaint (e.g., get_value_set is now read_value_set, and a new get_value_set has been introduced)

smowton commented:
Ah crap, sure, that was because the 4-operand and 3-operand versions of `get_value_set` were needlessly overloads, which led to awkwardness when subclassing `value_sett` as the 3-op version was hidden by overriding the 4-op one. I can make an alias instead. FWIW there isn't a new get_value_set, only the old 4-op version probably showing through where existing code expects the 3-op one. Again the cause was thinking value_set_domaint was a "private" implementation artefact of value_set_analysist.

In any case this is all now reverted per @kroening.

smowton commented:
@polgreen made a replacement to this PR here: https://github.com/diffblue/cbmc/pull/1499

Since this broke your stuff the last time around, would you mind checking whether you can merge that PR into your cbmc branch without causing undue mayhem?

polgreen commented:
That's great, thanks!
